wikitools/code_block_parser.py

```diff
@@ -0,0 +1,117 @@
+import typing
+
+
+class CodeTag(typing.NamedTuple):
+    start: int
+    len: int
+
+
+class CodeBlock(typing.NamedTuple):
+    """
+    A Markdown code block (monospace font + gray backround). Could be one of the following:
+        - Inline: `text` or `` text `` (the latter is used for snippets containing `)
+        - Multiline: ``` followed by anything else until the next ``` block
+
+    Examples:
+
+        ```test test test```
+        ^ start            ^ end
+
+        `this is a very long phrase`
+        ^ start                    ^ end
+
+    Similarly to the HTML comments, links in code blocks are not accounted for.
+    """
+
+    start: int  # 0-based position of the first character of the block's opening tag
+    end: int  # 0-based position of the last character of the block's closing tag
+    tag_len: int
+
+    @property
+    def is_multiline(self):
+        return self.end == -1 and self.tag_len == 3
+
+    def contains(self, other: 'CodeBlock'):
+        return (
+            self.start < other.start and
+            self.end > other.end and
+            self.tag_len > other.tag_len
+        )
+
+
+class CodeBlockParser:
+    def __init__(self):
+        self.__in_multiline = False
+
+    @property
+    def in_multiline(self) -> bool:
+        return self.__in_multiline
+
+    def parse(self, line: str) -> typing.List[CodeBlock]:
+        blocks: typing.List[CodeBlock] = []
+
+        tag_stack = []
+        if self.__in_multiline:
+            tag_stack.append(CodeTag(start=-1, len=3))
```

Comment:
this assumes a multiline code block always starts with a ` ``` `, when it can also start with ` ```` ` or longer

for example

````markdown
```
code block in a code block
```
````

no need to support it though since the wiki never uses it \:)

Reply:
yeah that's a deliberate decision